File: 1 - metodosBusca/pacoteBusca_JomiHubner/src/busca/BuscaBidirecional.java
package busca;

import java.util.LinkedList;
import java.util.List;

/**
 * Busca Bi-direcional.
 *
 *  @author Jomi Fred Hübner
 */
public class BuscaBidirecional extends Busca {
    
    /** busca sem mostrar status */
    public BuscaBidirecional() {
    }
    
    /** busca mostrando status */
    public BuscaBidirecional(MostraStatusConsole ms) {
        super(ms);
    }

    /**
     *
     * Busca a solução por busca em Bi-direcional.
     *                              -------------
     */
    public Nodo busca(Estado inicial, Estado meta) {
        status.inicia();
        usarFechado = false; // tem que usar poda so por ascendencia! nao pode usar fechados
        
        // TODO: ter duas tabelas de fechados, uma para cada lado!
        
        // TODO: usar tree para ter ordem (g) e ser rapido de achar
        List<Nodo> abertosCima  = new LinkedList<Nodo>();
        List<Nodo> abertosBaixo = new LinkedList<Nodo>();
        
        abertosCima.add(new Nodo(inicial, null));
        Nodo nodoMeta = new Nodo(meta, null);
        abertosBaixo.add(nodoMeta);
        
        while (!parar && abertosCima.size() > 0 && abertosBaixo.size() > 0) {
            
            // incrementa em cima
            //
            Nodo n = abertosCima.remove(0);
            status.explorando(n, abertosCima.size()+abertosBaixo.size());
            // ve se tem n na borda da árvore de baixo
            int io = abertosBaixo.indexOf(n); 
            if (io >= 0) {
                Nodo nb = abertosBaixo.get(io);
                nb.invertePaternidade();
                nb.pai = n.pai;
                nodoMeta.setProfundidade();
                status.termina(true);
                return nodoMeta;
            }
            abertosCima.addAll( sucessores(n) );
            
            // incrementa para baixo
            //
            n = abertosBaixo.remove(0);
            status.explorando(n, abertosCima.size()+abertosBaixo.size());
            // ve se tem n na borda da árvore de cima
            io = abertosCima.indexOf(n);
            if (io >= 0) {
                Nodo nc = abertosCima.get(io);
                n.invertePaternidade();
                n.pai = nc.pai;
                nodoMeta.setProfundidade();
                status.termina(true);
                return nodoMeta;
            }
            abertosBaixo.addAll( antecessores(n) );
            
        }
        status.termina(false);
        return null;
    }

    
    public Nodo busca(Estado inicial) throws Exception {
    	throw new Exception("Esta classe não implementa a busca com um único parâmetro"); 
    }
    
    public String toString() {
    	return "BBD - busca bi-direcional";
    }
}

File: 1 - metodosBusca/python/app/core/busca/busca_bidirecional.py
from collections import deque
from .busca import Busca
from .nodo import Nodo

class BuscaBidirecional(Busca):
    def __init__(self, mostra_status_console=None):
        super().__init__(mostra_status_console)

    def busca_with_meta(self, inicial, meta):
        self.status.inicia()
        self.usar_fechado = False  # tem que usar poda so por ascendencia! nao pode usar fechados

        # TODO: ter duas tabelas de fechados, uma para cada lado!
        
        # TODO: usar tree para ter ordem (g) e ser rapido de achar
        abertos_cima = deque([Nodo(inicial, None)])
        nodo_meta = Nodo(meta, None)
        abertos_baixo = deque([nodo_meta])
        
        while not self.parar and abertos_cima and abertos_baixo:
            
            # incrementa em cima
            n = abertos_cima.popleft()
            self.status.explorando(n, len(abertos_cima) + len(abertos_baixo))
            # ve se tem n na borda da árvore de baixo
            if n in abertos_baixo:
                nb = abertos_baixo[abertos_baixo.index(n)]
                nb.inverte_paternidade()
                nb.pai = n.pai
                nodo_meta.set_profundidade()
                self.status.termina(True)
                return nodo_meta
            abertos_cima.extend(self.sucessores(n))
            
            # incrementa para baixo
            n = abertos_baixo.popleft()
            self.status.explorando(n, len(abertos_cima) + len(abertos_baixo))
            # ve se tem n na borda da árvore de cima
            if n in abertos_cima:
                nc = abertos_cima[abertos_cima.index(n)]
                n.inverte_paternidade()
                n.pai = nc.pai
                nodo_meta.set_profundidade()
                self.status.termina(True)
                return nodo_meta
            abertos_baixo.extend(self.antecessores(n))
        
        self.status.termina(False)
        return None

    def busca(self, inicial):
        raise Exception("Esta classe não implementa a busca com um único parâmetro")

    def __str__(self):
        return "BBD - busca bi-direcional"